File: java/Advent02.java
import java.util.*;
import java.util.stream.*;

public class Advent02 extends Advent {

  int pc = 0;
  int[] nums;

  public Advent02() {
    super(2);
  }

  @Override
  protected void parseInput() {
    nums = Arrays.stream(input.get(0).split(","))
      .mapToInt(Integer::valueOf)
      .toArray();
    pc = 0;
  }

  @Override
  protected Object part1() {
    return runProgram(12, 2);
  }

  @Override
  protected Object part2() {
    for (int noun = 0; noun < 100; noun++) {
      for (int verb = 0; verb < 100; verb++) {
        if (runProgram(noun, verb) == 19690720) {
          return 100*noun + verb;
        }
      }
    }
    return null;
  }

  private int runProgram(int a, int b) {
    parseInput();
    nums[1] = a;
    nums[2] = b;
    runProgram();
    return nums[0];
  }

  private void runProgram() {
    int op1, op2, target;
    while (true) {
      switch (nums[pc]) {
        case 1:
          op1 = nums[nums[pc + 1]];
          op2 = nums[nums[pc + 2]];
          target = nums[pc + 3];
          nums[target] = op1 + op2;
          break;
        case 2:
          op1 = nums[nums[pc + 1]];
          op2 = nums[nums[pc + 2]];
          target = nums[pc + 3];
          nums[target] = op1 * op2;
          break;
        case 99:
          return;
      }
      pc += 4;
    }
  }
}

File: python/advent_02.py
from advent_lib import *
import itertools as iter

input = to_nums(lines(2)[0].split(","))

def eval(a, b):
    nums = input.copy()
    nums[1:3] = (a, b)
    pos = 0
    (op, a, b, c) = nums[pos:pos+4]
    while op != 99:
        if   op == 1: nums[c] = nums[a] + nums[b]
        elif op == 2: nums[c] = nums[a] * nums[b]
        pos += 4
        (op, a, b, c) = nums[pos:pos+4]
    return nums[0]

print("Part 1: ", eval(12, 2))

for noun, verb in iter.product(range(100), range(100)):
    if eval(noun, verb) == 19690720:
        print("Part 2: ", noun*100 + verb)
        break
